### PR TITLE
CODETOOLS-7903008: JMH: Support incremental annotation processing for Gradle Java plugin

### PR DESCRIPTION
**Pull request by twicksell**
This change adds support for Gradle's [incremental annotation processing](https://docs.gradle.org/current/userguide/java_plugin.html#sec:incremental_annotation_processing). Without this support, developers using the JMH annotation processor in their Gradle projects will see their compilation time increased due to Gradle invalidating its incremental compilation cache.  

/covered

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Issue
 * [CODETOOLS-7903008](https://bugs.openjdk.java.net/browse/CODETOOLS-7903008): JMH: Support incremental annotation processing for Gradle Java plugin

### Reviewers
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jmh pull/43/head:pull/43` \
`$ git checkout pull/43`

Update a local copy of the PR: \
`$ git checkout pull/43` \
`$ git pull https://git.openjdk.java.net/jmh pull/43/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 43`

View PR using the GUI difftool: \
`$ git pr show -t 43`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jmh/pull/43.diff">https://git.openjdk.java.net/jmh/pull/43.diff</a>

</details>

### CODE REVIEW

**Comment by robilad** [comment]
Hi, please send me an e-mail at Dalibor.topic@oracle.com so that I can mark your account as verified.

**Review by shipilev — APPROVED** [review]
Although there is no first-class support for Gradle in JMH, adding this piece of metadata should not hurt. I think JMH annotation processors satisfy the requirements of Gradle Java plugin.

How did you test it?

**Comment by twicksell** [comment]
Unfortunately I had to test this manually by performing a `mvn install` of my branch and using that with a couple of my own gradle projects.  If you run gradle with `--info` and you will see:
```
Full recompilation is required because org.openjdk.jmh.generators.BenchmarkProcessor is not incremental. Analysis took 0.085 secs.
```

After applying this change, you will see:
```
Created classpath snapshot for incremental compilation in 0.391 secs. 1153 duplicate classes found in classpath (see all with --debug).
```

I don't have a great way to add a test case to JMH for this, as I suspect you wouldn't want all of those gradle specifics as part of your build. But open to suggestions here.

**Comment by sfc-gh-twicksell** [comment]
Hold off on merging, I've found a place that wasn't using Filer to write state. `StateObjectHandler.writeStateOverrides` seems problematic. Looking to see if it can be fixed.

**Comment by twicksell** [comment]
Please hold off on merging. After running through some more complex benchmark examples, I've found a place where we are not using Filer to write classes. `StateObjectHandler.writeStateOverrides` seems problematic. Looking to see if I can fix that.

**Comment by shipilev** [comment]
> Please hold off on merging. After running through some more complex benchmark examples, I've found a place where we are not using Filer to write classes. `StateObjectHandler.writeStateOverrides` seems problematic. Looking to see if I can fix that.

I don't see a problem there? `StateObjectHandler.writeStateOverrides` delegates to `GeneratorDestination.newClass`, which in calls `APGeneratorDestination.newClass` in annotation processing mode, which uses `Filer`.

**Comment by twicksell** [comment]
Sorry to be a bother, but would you mind having another look at this one @shipilev?

**Comment by shipilev** [comment]
> Sorry to be a bother, but would you mind having another look at this one @shipilev?

Hi! Sorry for delay. This looks marginally okay, although I was getting a bit nervous about implementing new stuff to support (unsupported) Gradle builds. But I think this is okay. 

Now that it affects core code, we have to figure out a bit of testing. Please merge from the `master`, and see if https://github.com/twicksell/jmh/actions have workflows enabled for this PR. Those should be green.

**Comment by sfc-gh-twicksell** [comment]
Thanks @shipilev. I've updated my PR to bring it up to date with master. It looks like I need some kind of approval for the github action to run. As for the test plan, I'm open to anything you'd like to see. Really appreciate your help in getting this merged, even though Gradle is not explicitly supported.

**Comment by twicksell** [comment]
Thanks @shipilev. I've updated my PR to bring it up to date with master. It looks like I need some kind of approval for the github action to run. As for the test plan, I'm open to anything you'd like to see. Really appreciate your help in getting this merged, even though Gradle is not explicitly supported.

**Comment by shipilev** [comment]
> Thanks @shipilev. I've updated my PR to bring it up to date with master. It looks like I need some kind of approval for the github action to run. As for the test plan, I'm open to anything you'd like to see. Really appreciate your help in getting this merged, even though Gradle is not explicitly supported.

I see testing is green, which is good.

I have a question, though. Reading the Gradle docs, I see:

> "Isolating" processors have the following limitations:
> 
>  - They must provide exactly one originating element for each file generated with the Filer API. If zero or many originating elements are provided, Gradle will recompile all source files.

Does it make sense to pass varargs for originating elements through internal JMH APIs then? It would probably be cleaner to just pass one originating element, so we don't accidentally pass more than one, and thus make Gradle suffer?

**Comment by twicksell** [comment]
> Does it make sense to pass varargs for originating elements through internal JMH APIs then? It would probably be cleaner to just pass one originating element, so we don't accidentally pass more than one, and thus make Gradle suffer?

You're correct. I chose varargs because I didn't want to impose that Gradle-specific detail onto JMH, particularly because `GeneratorDestination` is a fairly generic concept. I'm happy to make this change if you're willing to accept it.

**Comment by shipilev** [comment]
> You're correct. I chose varargs because I didn't want to impose that Gradle-specific detail onto JMH, particularly because `GeneratorDestination` is a fairly generic concept. Varargs also helped a bit with backwards compatibility, as the argument can be ignored. Otherwise I'll be adding an overloaded method to `GeneratorDestination`.
> 
> I'm happy to make this change if you're willing to accept it.

Let's do a single parameter. This would mechanically force the code to follow what Gradle wants.

If I read the code right, all uses of `GeneratorDestination.newClass` are modified by your patch already, so you don't need an overload. Backwards compatibility is not a problem here: users are expected to match the generator versions precisely.

File: jmh-core-ct/src/test/java/org/openjdk/jmh/ct/InMemoryGeneratorDestination.java
Comment (shipilev) [inline jmh-core-ct/src/test/java/org/openjdk/jmh/ct/InMemoryGeneratorDestination.java:67]: Here and later, should be singular `originatingClassName`.

File: jmh-core-ct/src/test/java/org/openjdk/jmh/ct/InMemoryGeneratorDestination.java
Comment (shipilev) [inline jmh-core-ct/src/test/java/org/openjdk/jmh/ct/InMemoryGeneratorDestination.java:64]: Here and later. I don't think you need this override. Just pass `null` where needed. This would save us from the accidental passing of `null` where originating element is actually available.

File: jmh-core/src/main/java/org/openjdk/jmh/generators/core/GeneratorDestination.java
Comment (shipilev) [inline jmh-core/src/main/java/org/openjdk/jmh/generators/core/GeneratorDestination.java:61]: `class name`, singular.

File: jmh-generator-annprocess/src/main/java/org/openjdk/jmh/generators/annotations/APGeneratorDestinaton.java
Comment (shipilev) [inline jmh-generator-annprocess/src/main/java/org/openjdk/jmh/generators/annotations/APGeneratorDestinaton.java:63]:
```suggestion
        if (originatingClassNames != null) {
```

File: jmh-generator-annprocess/src/main/java/org/openjdk/jmh/generators/annotations/APGeneratorDestinaton.java
Comment (shipilev) [inline jmh-generator-annprocess/src/main/java/org/openjdk/jmh/generators/annotations/APGeneratorDestinaton.java:66]: I don't think this even compiles ;)

**Review by shipilev — CHANGES_REQUESTED** [review]

File: jmh-core-ct/src/test/java/org/openjdk/jmh/ct/InMemoryGeneratorDestination.java
Comment (sfc-gh-twicksell) [reply jmh-core-ct/src/test/java/org/openjdk/jmh/ct/InMemoryGeneratorDestination.java:64]: Fair enough. Sorry about that, I'm just used to being defensive about making backwards incompatible changes. 😄

File: jmh-generator-annprocess/src/main/java/org/openjdk/jmh/generators/annotations/APGeneratorDestinaton.java
Comment (twicksell) [reply jmh-generator-annprocess/src/main/java/org/openjdk/jmh/generators/annotations/APGeneratorDestinaton.java:66]: That's my fault for trying to make a "quick fix" while traveling for the holidays. Sorry about that, correct change is pushed. I'm running through my manual verifications in the meanwhile just to make sure nothing funky changed.

File: jmh-generator-annprocess/src/main/java/org/openjdk/jmh/generators/annotations/APGeneratorDestinaton.java
Comment (shipilev) [reply jmh-generator-annprocess/src/main/java/org/openjdk/jmh/generators/annotations/APGeneratorDestinaton.java]: This is still a suggested change :)

File: jmh-generator-annprocess/src/main/java/org/openjdk/jmh/generators/annotations/APGeneratorDestinaton.java
Comment (shipilev) [inline jmh-generator-annprocess/src/main/java/org/openjdk/jmh/generators/annotations/APGeneratorDestinaton.java:62]: Um, this branch is the wrong way. `originatingClassName` is guaranteed to be `null` at `else` branch.

```suggestion
    @Override
    public Writer newClass(String className, String originatingClassName) throws IOException {
        Filer filer = processingEnv.getFiler();
        if (originatingClassName != null) {
            TypeElement originatingType = processingEnv.getElementUtils().getTypeElement(originatingClassName);
            return filer.createSourceFile(className, originatingType).openWriter();
        } else {
            return filer.createSourceFile(className).openWriter();
        }
    }
```

**Review by shipilev — CHANGES_REQUESTED** [review]
Still not quite right, sorry.

**Comment by shipilev** [comment]
FYI, tomorrow is the last working day this year. If you could complete this PR until then, we could probably even have a minor JMH release before the break.

**Review by shipilev — APPROVED** [review]
Assuming GitHub checks are clean, you are free to say `/integrate`.

**Pull request closed without merging**
